File: behavioral/state/vending_machine.py
from abc import ABC, abstractmethod


# State interface
class State(ABC):
    @abstractmethod
    def insert_coin(self, machine: "VendingMachine") -> None:
        pass

    @abstractmethod
    def press_button(self, machine: "VendingMachine") -> None:
        pass

    @abstractmethod
    def dispense(self, machine: "VendingMachine") -> None:
        pass


# Concrete States
class IdleState(State):
    def insert_coin(self, machine: "VendingMachine") -> None:
        print("Coin inserted")
        machine.set_state(machine.has_coin_state)

    def press_button(self, machine: "VendingMachine") -> None:
        print("Insert a coin first")

    def dispense(self, machine: "VendingMachine") -> None:
        print("Insert a coin first")


class HasCoinState(State):
    def insert_coin(self, machine: "VendingMachine") -> None:
        print("Coin already inserted")

    def press_button(self, machine: "VendingMachine") -> None:
        print("Button pressed")
        machine.set_state(machine.dispensing_state)

    def dispense(self, machine: "VendingMachine") -> None:
        print("Press the button first")


class DispensingState(State):
    def insert_coin(self, machine: "VendingMachine") -> None:
        print("Please wait, dispensing in progress")

    def press_button(self, machine: "VendingMachine") -> None:
        print("Already dispensing")

    def dispense(self, machine: "VendingMachine") -> None:
        if machine.count > 0:
            machine.count -= 1
            print("Item dispensed!")
        if machine.count == 0:
            print("Out of stock!")
            machine.set_state(machine.out_of_stock_state)
        else:
            machine.set_state(machine.idle_state)


class OutOfStockState(State):
    def insert_coin(self, machine: "VendingMachine") -> None:
        print("Machine is out of stock")

    def press_button(self, machine: "VendingMachine") -> None:
        print("Machine is out of stock")

    def dispense(self, machine: "VendingMachine") -> None:
        print("Machine is out of stock")


# Context
class VendingMachine:
    def __init__(self, count: int):
        self.idle_state         = IdleState()
        self.has_coin_state     = HasCoinState()
        self.dispensing_state   = DispensingState()
        self.out_of_stock_state = OutOfStockState()

        self.count = count
        self._state: State = self.idle_state if count > 0 else self.out_of_stock_state

    def set_state(self, state: State) -> None:
        self._state = state

    def insert_coin(self) -> None:
        self._state.insert_coin(self)

    def press_button(self) -> None:
        self._state.press_button(self)

    def dispense(self) -> None:
        self._state.dispense(self)


# Example usage
if __name__ == "__main__":
    machine = VendingMachine(count=2)

    print("-- Normal purchase --")
    machine.insert_coin()
    machine.press_button()
    machine.dispense()

    print("\n-- Invalid actions --")
    machine.press_button()
    machine.dispense()

    print("\n-- Last item --")
    machine.insert_coin()
    machine.press_button()
    machine.dispense()

    print("\n-- Out of stock --")
    machine.insert_coin()
    

File: behavioral/state/VendingMachine.java
// State interface
interface State {
    void insertCoin(VendingMachine machine);
    void pressButton(VendingMachine machine);
    void dispense(VendingMachine machine);
}


// Concrete States
class IdleState implements State {
    @Override
    public void insertCoin(VendingMachine machine) {
        System.out.println("Coin inserted");
        machine.setState(machine.hasCoinState);
    }

    @Override
    public void pressButton(VendingMachine machine) {
        System.out.println("Insert a coin first");
    }

    @Override
    public void dispense(VendingMachine machine) {
        System.out.println("Insert a coin first");
    }
}


class HasCoinState implements State {
    @Override
    public void insertCoin(VendingMachine machine) {
        System.out.println("Coin already inserted");
    }

    @Override
    public void pressButton(VendingMachine machine) {
        System.out.println("Button pressed");
        machine.setState(machine.dispensingState);
    }

    @Override
    public void dispense(VendingMachine machine) {
        System.out.println("Press the button first");
    }
}


class DispensingState implements State {
    @Override
    public void insertCoin(VendingMachine machine) {
        System.out.println("Please wait, dispensing in progress");
    }

    @Override
    public void pressButton(VendingMachine machine) {
        System.out.println("Already dispensing");
    }

    @Override
    public void dispense(VendingMachine machine) {
        if (machine.count > 0) {
            machine.count--;
            System.out.println("Item dispensed!");
        }
        if (machine.count == 0) {
            System.out.println("Out of stock!");
            machine.setState(machine.outOfStockState);
        } else {
            machine.setState(machine.idleState);
        }
    }
}


class OutOfStockState implements State {
    @Override
    public void insertCoin(VendingMachine machine) {
        System.out.println("Machine is out of stock");
    }

    @Override
    public void pressButton(VendingMachine machine) {
        System.out.println("Machine is out of stock");
    }

    @Override
    public void dispense(VendingMachine machine) {
        System.out.println("Machine is out of stock");
    }
}


// Context
public class VendingMachine {
    public final State idleState       = new IdleState();
    public final State hasCoinState    = new HasCoinState();
    public final State dispensingState = new DispensingState();
    public final State outOfStockState = new OutOfStockState();

    public int count;
    private State state;

    public VendingMachine(int count) {
        this.count = count;
        this.state = count > 0 ? idleState : outOfStockState;
    }

    public void setState(State state) { this.state = state; }

    public void insertCoin()  { state.insertCoin(this); }
    public void pressButton() { state.pressButton(this); }
    public void dispense()    { state.dispense(this); }

    public static void main(String[] args) {
        VendingMachine machine = new VendingMachine(2);

        System.out.println("-- Normal purchase --");
        machine.insertCoin();
        machine.pressButton();
        machine.dispense();

        System.out.println("\n-- Invalid actions --");
        machine.pressButton();
        machine.dispense();

        System.out.println("\n-- Last item --");
        machine.insertCoin();
        machine.pressButton();
        machine.dispense();

        System.out.println("\n-- Out of stock --");
        machine.insertCoin();
    }
}